runtime/fail_nat.c

```diff
@@ -217,6 +222,11 @@ void caml_array_bound_error(void)
 
 void caml_array_bound_error_asm(void)
 {
+#if defined(WITH_THREAD_SANITIZER)
+  char* exception_pointer = (char*)Caml_state->c_stack;
+  caml_tsan_exn_func_exit_c(exception_pointer);
```

Comment:
Why are you using the `exit_c` variant here? I would in general assume that we are running on an OCaml stack rather than a C stack when calling this function.

Reply:
When this function is run, we are on a C stack: it is called with `caml_c_call`.